### PR TITLE
fix(opencode): remove invalid top-level theme config

### LINKED ISSUE #497: fix(opencode): remove invalid top-level theme config
### Pre-flight Checklist

- [x] I have searched [existing issues](https://github.com/Gentleman-Programming/gentle-ai/issues) and this is not a duplicate
- [x] I understand that PRs will be rejected if the linked issue does not have `status:approved`

### 📝 Bug Description

Gentle AI writes this top-level key to OpenCode's `opencode.json`:

```json
"theme": "gentleman-kanagawa"
```

OpenCode 1.14.46 rejects that key as invalid:

```txt
Unrecognized key: theme
```

This breaks runtime config updates, including runtime profile activation, because OpenCode validates the full config before accepting the update.

### 🔄 Steps to Reproduce

1. Enable/sync the Gentle AI theme for OpenCode.
2. Confirm `~/.config/opencode/opencode.json` contains top-level `"theme": "gentleman-kanagawa"`.
3. Trigger any OpenCode runtime config update, for example profile activation.
4. OpenCode rejects the update with `ConfigInvalidError`.

### ✅ Expected Behavior

Gentle AI should not write OpenCode config keys that OpenCode's current schema rejects.

### ❌ Actual Behavior

Runtime config updates fail until the top-level `theme` key is removed from `opencode.json`.

### Gentle AI Version

gentle-ai 1.26.6

### Operating System

Linux (Fedora/RHEL)

### AI Agent / Client

OpenCode

### 📋 Affected Area

Installation Pipeline

### 💡 Logs / Error Output

```shell
ConfigInvalidError: Unrecognized key: theme
```

### Additional Context

Environment:

- OpenCode: 1.14.46
- Engram: 1.15.10
- Gentle AI: 1.26.6

Code pointer:

```go
// internal/components/theme/inject.go
var themeOverlayJSON = []byte("{\n  \"theme\": \"gentleman-kanagawa\"\n}\n")
```

Removing `"theme": "gentleman-kanagawa"` from `opencode.json` makes profile activation work again.

### PR DESCRIPTION
**Pull request by LCubero**
## 🔗 Linked Issue

Closes #497

---

## 🏷️ PR Type

- [x] `type:bug` — Bug fix (non-breaking change that fixes an issue)
- [ ] `type:feature` — New feature (non-breaking change that adds functionality)
- [ ] `type:docs` — Documentation only
- [ ] `type:refactor` — Code refactoring (no functional changes)
- [ ] `type:chore` — Build, CI, or tooling changes
- [ ] `type:breaking-change` — Breaking change

---

## 📝 Summary

- Prevents Gentle AI from writing the unsupported top-level `theme` key into OpenCode's `opencode.json`.
- Migrates existing literal root `theme` entries while preserving nested settings, file permissions, and covered JSONC comments and formatting.
- Rejects malformed JSONC without modifying the source file and guarantees scanner progress on malformed internal input.

---

## 📂 Changes

| File / Area | What Changed |
|-------------|--------------|
| Agent capabilities | Adds explicit theme-injection and legacy-migration adapter capabilities |
| OpenCode adapter | Skips new theme injection and removes legacy literal root `theme` entries |
| JSONC removal | Removes only the targeted root member without reserializing unrelated content |
| Scanner safety | Rejects malformed token splicing, guards no-progress scans, and preserves LF/CRLF boundaries |
| Theme workflow | Runs migration before applying the adapter's current injection policy |
| CLI paths | Includes existing OpenCode settings in backup and verification handling |
| Catalog and docs | Describes Theme as a setting for supported agents rather than an OpenCode theme |
| Tests and E2E | Adds regression coverage for migration, comments, malformed input, idempotency, modes, scanners, and platform behavior |

---

## 🧪 Test Plan

**Focused tests**

```bash
go test ./internal/components/filemerge/... ./internal/components/theme/... -count=1
```

**Full unit suite**

```bash
go test ./... -count=1
```

**Docker E2E**

```bash
cd e2e && ./docker-test.sh
```

Results:

- [x] Focused `filemerge` and `theme` tests pass locally
- [x] Shell syntax and focused JSONC fixtures pass locally
- [x] Go Format passes in GitHub Actions
- [x] Unit Tests pass in GitHub Actions
- [x] Windows Runtime passes in GitHub Actions
- [x] E2E Tests pass on Ubuntu, Arch, and Fedora
- [ ] Full Docker Tier 2 was not run locally and is not claimed

---

## 🤖 Automated Checks

| Check | Status | Description |
|-------|--------|-------------|
| Check PR Cognitive Load | BLOCKED | Requires maintainer-applied `size:exception` |
| Check Issue Reference | PASS | `Closes #497` |
| Check Issue Has `status:approved` | PASS | Issue #497 is approved |
| Check PR Has `type:*` Label | BLOCKED | Requires maintainer-applied `type:bug` |
| Go Format | PASS | Passed on the final candidate |
| Unit Tests | PASS | Passed on the final candidate |
| Windows Runtime | PASS | Passed on the final candidate |
| E2E Tests | PASS | Passed on Ubuntu, Arch, and Fedora |
| CodeRabbit | PASS | Review completed on the final head |

---

## ✅ Contributor Checklist

- [x] PR is linked to an issue with `status:approved`
- [x] I have requested maintainer-applied `size:exception` with the rationale documented below
- [ ] I have added the appropriate `type:*` label to this PR
- [x] Unit tests pass (`go test ./...`)
- [x] E2E tests pass (`cd e2e && ./docker-test.sh`)
- [x] I have updated documentation where behavior changed
- [x] My commits follow Conventional Commits format
- [x] My commits do not include `Co-Authored-By` trailers

---

### Suggested review path

1. OpenCode adapter opt-out and legacy migration.
2. Byte-preserving root-key removal and fail-closed JSONC validation.
3. Backup and post-apply verification path handling.
4. Regression tests and Docker E2E coverage.

### Review-size rationale

Requesting `size:exception` for one cohesive migration and safety boundary.

| Measure | Lines |
|---------|------:|
| Total change | 750 |
| Tests, E2E, goldens, and docs | 463 |
| Production implementation | 287 |
| Standard budget exceeded by | 350 |

A chained PR was considered. It would separate legacy migration from JSONC safety and the regression proof required to review it confidently, leaving incomplete intermediate states.

Please apply the maintainer-managed `size:exception` label.

### Label permissions

I have read-only permission on the upstream repository and cannot apply PR labels.

Please add:

- `type:bug`
- `size:exception`

### Preservation guarantees

- Removes Gentle AI's legacy literal root `theme` member.
- Preserves nested `theme` members.
- Preserves the existing file mode.
- Preserves covered comments, whitespace, ordering, and LF/CRLF formatting.
- Leaves malformed files byte-identical.
- Repeated migration is idempotent.
- Scanner helpers terminate when comment parsing makes no progress.

### Intentionally out of scope

- Escaped semantic spellings of the root `theme` key
- Writing a theme into `tui.json`
- Shared `json_merge.go` behavior
- Claude issue #896 / PR #1061
- Kilocode support

### Review follow-up

- Commit `919617e` addressed the isolated-slash scanner finding, and CodeRabbit marked the thread as resolved.
- Commit `2e04de2` made the Windows permission assertion platform-aware, accumulated migration and injection results, deduplicated affected paths, and added root-only OpenCode theme validation.
- Commit `7f1f963` removed the dead Python test stub and preserves accumulated migration results when later theme injection fails.
- Commit `52aa4d2` aligns the idempotent OpenCode theme checks with root-only JSONC semantics.
- Fresh checks pass for Go Format, Unit Tests, Windows Runtime, and E2E on Ubuntu, Arch, and Fedora.

## Summary by CodeRabbit

* **Bug Fixes**
  * OpenCode no longer writes or expects a root-level `theme` key in `opencode.json`.
  * Legacy OpenCode theme settings are migrated by removing only the root `theme` key while preserving the rest of the file (including JSONC formatting).
  * Theme injection behavior is now adapter-aware and avoids missing-file/merge issues; repeated runs are idempotent.

* **Documentation**
  * Updated the Gentleman Kanagawa Theme description to clarify it’s for supported agents and is a no-op for OpenCode.

* **Tests**
  * Expanded unit and end-to-end coverage for JSONC handling, malformed inputs, and the absence of a root `theme` key.

### CODE REVIEW

**Review by Alan-TheGentleman — CHANGES_REQUESTED** [review]
The prevention path is solid, but I'm requesting changes because existing OpenCode installations can retain the invalid top-level `theme` key. `Inject` now returns before reading the settings file, and post-apply verification skips that file, so an install can report success while profile activation remains broken. Please migrate existing configurations by removing only the stale top-level `theme` key, preserve every other setting, add a regression test that starts from an existing `opencode.json`, and add the required `type:*` label. Push the update and I'll re-review.

**Review by Alan-TheGentleman — CHANGES_REQUESTED** [review]
The migration now addresses the original failure, but I'm keeping changes requested because the remover still does not preserve unrelated content or fail closed for every malformed file. When `theme` is the first top-level member, the removal range also deletes preceding JSONC comments; an unterminated trailing block comment can also be normalized away before validation, allowing a malformed file to be rewritten instead of left untouched. Please add regression tests for both cases and adjust the parser and removal boundaries so unrelated comments are preserved and every malformed input returns without mutation. The required `type:*` label is also still missing, and its policy check is failing. Push the update and I'll re-review.

**Comment by dnlrsls** [comment]
Please make the permission-preservation contract portable on Windows, or scope the POSIX mode assertion to platforms where those bits are meaningful. `go test ./internal/components/theme -run '^TestInjectRemovesLegacyOpenCodeThemeOnly$' -count=3` fails deterministically on Windows/amd64 because the rewritten file reports mode `0666` while the new test requires `0600`.

This test is candidate-only and does not exist in frozen base tree `c7e46beaf7bba4bae5a12531c853b6a7690b7649`. The other deterministic full-suite failures reproduced on that frozen base and are not attributed to this PR; one MCP timeout was load-sensitive. Please rerun the focused test after adjusting the contract or assertion.

## Validated patch

The minimal test-only fix is to keep the permission assertion on platforms where POSIX mode bits are meaningful:

```diff
diff --git a/internal/components/theme/inject_test.go b/internal/components/theme/inject_test.go
index 1e47f1d..84851fe 100644
--- a/internal/components/theme/inject_test.go
+++ b/internal/components/theme/inject_test.go
@@ -4,6 +4,7 @@ import (
 	"encoding/json"
 	"os"
 	"path/filepath"
+	"runtime"
 	"testing"
 
 	"github.com/gentleman-programming/gentle-ai/internal/agents"
@@ -145,8 +146,10 @@ func TestInjectRemovesLegacyOpenCodeThemeOnly(t *testing.T) {
 	if err != nil {
 		t.Fatalf("Stat(settings) error = %v", err)
 	}
-	if got := info.Mode().Perm(); got != 0o600 {
-		t.Fatalf("settings mode = %o, want preserved 600", got)
+	if runtime.GOOS != "windows" {
+		if got := info.Mode().Perm(); got != 0o600 {
+			t.Fatalf("settings mode = %o, want preserved 600", got)
+		}
 	}
 
 	second, err := Inject(home, opencodeAdapter())
```

Validated against commit `919617e437d061ae62bfbd7fdebd9681367ecf8b` on Windows/amd64:

```text
$ go test ./internal/components/theme -run '^TestInjectRemovesLegacyOpenCodeThemeOnly$' -count=3
ok  	github.com/gentleman-programming/gentle-ai/internal/components/theme	1.009s

$ go test ./internal/components/theme -count=1
ok  	github.com/gentleman-programming/gentle-ai/internal/components/theme	0.877s

$ GOOS=linux GOARCH=amd64 GOPROXY=off go test -c ./internal/components/theme
PASS (compile-only, no output)
```

**Comment by LCubero** [comment]
@coderabbitai review

Updated in `7f1f963` with both latest review follow-ups:

- removed the unused Python stub from the Node-unavailable E2E simulation
- preserved the accumulated migration result when later theme injection fails
- added deterministic regression coverage for the partial-result contract

Fresh checks pass for Go Format, Unit Tests, Windows Runtime, and E2E on Ubuntu, Arch, and Fedora.

**Comment by LCubero** [comment]
@Alan-TheGentleman PR #1287 is ready for re-review at `52aa4d2`.

All reviewer and CodeRabbit follow-ups are addressed. Fresh checks pass for Go Format, Unit Tests, Windows Runtime, and E2E on Ubuntu, Arch, and Fedora.

Please apply `type:bug` and `size:exception`. The PR body includes the final 750-line review-size breakdown and the rationale for keeping the migration and its safety proof together.

**Comment by dnlrsls** [comment]
Closing without merge: the maintainer confirmed that #497 is fixed on current main by commit `4955eb55` (https://github.com/Gentleman-Programming/gentle-ai/issues/497#issuecomment-5058482176). This PR is therefore superseded for its linked issue.
